core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -303,15 +311,25 @@ class LogSegment(val log: FileRecords,
     // We may need to reload the max timestamp after truncation.
     if (maxTimestampSoFar >= 0)
       loadLargestTimestamp(readToLogEnd = true)
+    _nextOffset = math.max(baseOffset, math.min(_nextOffset, offset))
     bytesTruncated
   }
 
   /**
-   * Calculate the offset that would be used for the next message to be append to this segment.
+   * The offset that would be used for the next message to be appended to this segment
+   */
+  def nextOffset(): Long = {
+    if (_nextOffset == INVALID_NEXT_OFFSET)
+      _nextOffset = nextOffsetFromLog()
```

Comment:
Before this patch, we have nextOffset(), which computes the next offset by scanning the last portion of the log. This could be expensive, but is only used during the initial log loading. I was suggesting that we could just keep the logic as it is. in deleteLogStartOffsetBreachedSegments(), we always use the baseOffset of the next segment to check if a segment should be deleted. If a segment doesn't have the next segment, we know it's the active segment and won't be deleted. This way, we never need to call nextOffset() in deleteLogStartOffsetBreachedSegments().